File: hackerearth/cpp/number-of-ways.cpp
/**
 * https://www.hackerearth.com/practice/algorithms/dynamic-programming/2-dimensional/practice-problems/algorithm/rook-path-142e55ee/
 **/
#include<bits/stdc++.h>
#define lli long long int
#define vi vector<Triplet>
#define vvi vector<vi>
#define MOD 1000000007
using namespace std;

class Triplet{
    public:
        lli num_w; // number of ways to reach this cell
        lli sum_l; // sum of number of ways to reach the cells to the left (till an obstacle) of the current cell.
        lli sum_u; // sum of number of ways to reach the cells to the top (till an obstacle) of the current cell.
        void set(lli w, lli l, lli u) {
            num_w = w;
            sum_l = l;
            sum_u = u;
        }
};

void solve_test_case() {
    int n, m;
    cin>>n>>m;
    vector<string> grid(n);
    for(int i=0; i<n; i++)
        cin>>grid[i];
    
    vvi dp(n, vi(m));
    for(int i=0;i<n;i++){
        for(int j=0;j<m;j++) {
            Triplet &ct = dp[i][j];
            if(i==0 && j==0) {
                ct.set(1, 1, 1);
                continue;
            }
            
            if(grid[i][j] == '*') {
                ct.set(0, 0, 0);
                continue;
            }

            lli sum_l = 0;
            lli sum_u = 0;
            if(j!=0) {
                sum_l = dp[i][j-1].sum_l;
            }

            if(i!=0) {
                sum_u = dp[i-1][j].sum_u;
            }

            lli num_ways = (sum_l + sum_u) % MOD;
            ct.set(num_ways, (num_ways + sum_l) % MOD, (sum_u + num_ways) % MOD);
        }
    }
    cout<<dp[n-1][m-1].num_w<<"\n";
}

int main() {
    int t;
    cin >> t;
    while (t--) {
        solve_test_case();
    }
}

File: hackerearth/golang/number-of-ways.go
/**
 * https://www.hackerearth.com/practice/algorithms/dynamic-programming/2-dimensional/practice-problems/algorithm/rook-path-142e55ee/
 **/
package main

import "fmt"

type tripletNumberOfWays struct {
	numW, sumL, sumU int64
}

var mod int64

func solveNumberOfWays() {
	var n, m int
	fmt.Scanf("%d %d", &n, &m)
	grid := make([]string, n)
	for i := 0; i < n; i++ {
		fmt.Scanf("%s", &grid[i])
	}

	dp := make([][]tripletNumberOfWays, n)
	for i := range dp {
		dp[i] = make([]tripletNumberOfWays, m)
	}

	for i := range dp {
		for j := range dp[i] {
			if i == 0 && j == 0 {
				dp[i][j] = tripletNumberOfWays{1, 1, 1}
				continue
			}

			if grid[i][j] == '*' {
				dp[i][j] = tripletNumberOfWays{0, 0, 0}
				continue
			}

			sumL := int64(0)
			sumU := int64(0)

			if j != 0 {
				sumL = dp[i][j-1].sumL
			}

			if i != 0 {
				sumU = dp[i-1][j].sumU
			}
			numW := (sumL + sumU) % mod
			dp[i][j] = tripletNumberOfWays{
				numW,
				(sumL + numW) % mod,
				(sumU + numW) % mod}
		}
	}
	fmt.Println(dp[n-1][m-1].numW)
}

func numberOfWays() {
	var t int
	mod = 1000000007
	fmt.Scanf("%d", &t)
	for t > 0 {
		solveNumberOfWays()
		t--
	}
}
